crates/ruff_linter/src/rules/flake8_cognitive_complexity/rules/function_is_too_cognitive_complex.rs

```diff
@@ -0,0 +1,281 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, identifier::Identifier, Expr, Stmt};
+
+/// ## What it does
+/// Checks for functions with a high cognitive complexity.
+///
+/// ## Why is this bad?
+/// Cognitive complexity is a metric based of a whitepaper by G. Ann Campbell.
+///
+///  > Cognitive Complexity breaks from the practice of using mathematical models to
```

Comment:
What's the licensing of Sonar's rule? The whitespaper has a footer mentioning "Copyright SonarSource S . A., 2023, Switzerland. All content is copyright protected". I don't know to what extend this applies to algorithms but something we should consider.

Reply:
https://github.com/SonarSource/sonar-python?tab=LGPL-3.0-1-ov-file
Looks like it is `LGPL`, so I guess??? Not completely sure.

Reply:
Nice. Yeah that looks okay. Also, code climate implements the same rule. So I think we should be good https://docs.codeclimate.com/docs/cognitive-complexity